From b92504706c1689f029701c5c1200ea0b4fc41c95 Mon Sep 17 00:00:00 2001
From: nemeth <nemeth@studlab1-u.dfki.uni-bremen.de>
Date: Mon, 16 Aug 2010 12:25:36 +0200
Subject: [PATCH 3/4] enable ipv6 in avahi-daemon.conf

---
 avahi-daemon/avahi-daemon.conf |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/avahi-daemon/avahi-daemon.conf b/avahi-daemon/avahi-daemon.conf
index c992842..a6286ad 100644
--- a/avahi-daemon/avahi-daemon.conf
+++ b/avahi-daemon/avahi-daemon.conf
@@ -23,7 +23,7 @@
 #domain-name=local
 browse-domains=0pointer.de, zeroconf.org
 use-ipv4=yes
-use-ipv6=no
+use-ipv6=yes
 #allow-interfaces=eth0
 #deny-interfaces=eth1
 #check-response-ttl=no
-- 
1.7.0.4

